### ISSUE TITLE
Add character detection blocks

### ISSUE THREAD

**Issue by filips123** [LLK/scratch-blocks#2018]
### Proposed Changes

Add blocks for detecting various information about a character. This should be "is digit", "is letter", "is special", "is uppercase" and "is lowercase".

### Reason for Changes

Scratch does only supports case-insensitive character comparison. Although there are already some possibilities to detect letter case, those are mostly some "hacks" and are not good for normal user. This blocks would make easier to detect information about the character and make projects that rely on text easier.

**Comment by filips123** [comment]
@chrisgarrity Do you think this additions would be good? If so, I can make PRs.

**Comment by Kenny2github** [comment]
@filips123 chrisgarrity is responsible for internationalization-related work (as far as I know), which has nothing to do with adding new blocks. Better to ping @thisandagain (just for acknowledgement, if nothing else), who is responsible for the project overall. (Don't ping him now, though - I just have.)

Additionally, even if this were to be implemented, please do not open any PRs unless the issue is labelled "help wanted" - there are design decisions that need to be made before new blocks are added.

There's also a few ambiguities that you've left unresolved:
- Scratch inputs can't be restricted to one character - what do these blocks report if given a multi-character string?
  - For example, for the "is uppercase" option: would that report true if any character was uppercase, or only if all of them were?
- These all have similar functions, so would it not be better to combine them into one `<is [ v]>` block?
- Specifically for "is special", what do you mean by that? Symbols, or ASCII [control characters](https://en.wikipedia.org/wiki/Control_character)?

**Comment by filips123** [comment]
> Scratch inputs can't be restricted to one character - what do these blocks report if given a multi-character string?

One of the ways to implement this is to [compare `toUpperCase` and `toLowerCase` with original string](https://stackoverflow.com/a/1077692/6523409). In this case, is uppercase would mean that whole string should be uppercase.

> These all have similar functions, so would it not be better to combine them into one <is [ v]> block?

Yes, that's what I thought but I forgot to describe it.

> Specifically for "is special", what do you mean by that? Symbols, or ASCII control characters?

Then there could be "is symbol" for symbols and "is control" for ASCII control characters.

**Comment by apple502j** [comment]
ASCII control characters cannot be inserted - or Scratch breaks. Also, what is letter, what is digit, what is uppercase and what is lowercase?

**Comment by filips123** [comment]
> ASCII control characters cannot be inserted - or Scratch breaks.

Ok. In this case, there should just be "is symbol" check.

> Also, what is letter, what is digit, what is uppercase and what is lowercase?

* The letter would be `char.toLowerCase() != char.toUpperCase()`. [It](https://stackoverflow.com/a/32567789/6523409) will work for most Latin, Greek, Armenian and Cyrillic scripts, but not for Chinese, Japanese, Arabic, Hebrew and most other scripts. It only works on one character, but can be fixed to iterate over the string and check each character separately. For other scripts, we would probably need to check if the character is in the correct Unicode range.

* Number would be `!isNaN(string * 1)`. This also works for a string. It also recognises HEX numbers as numbers.

* Lowercase and uppercase would be `string == string.toUpperCase()` and `string == string.toLowerCase()`. [This](https://stackoverflow.com/a/1077692/6523409) should work for same scripts as letter check from above. For other scripts that don't have uppercase and lowercase letters, it would always return `true`. In case if the string is given, whole string has to be uppercase or lowercase to match.

* Symbols would be other characters not matching any of the above conditions. The whole string should match.

Additionally, there could be <contains [ v]> block where it would return true if any of chars is letter/number/uppercase/lowercase/symbol.

**Comment by chrisgarrity** [comment]
@filips123 as previously mentioned, please don't submit PRs unless it gets labelled help-wanted. If we did decide that these blocks were something to try out, I expect that they would be added as an extension, not as core blocks. You can find out more about extensions in the VM, but that feature is still getting worked on: https://github.com/LLK/scratch-vm/blob/develop/docs/extensions.md

**Comment by filips123** [comment]
> please don't submit PRs unless it gets labelled help-wanted

Yes, sure. I just shared code to how how implementation could look like and details how could handle specific cases.

> I expect that they would be added as an extension, not as core blocks

I don't fully agree with adding them as an extension. They would just be two blocks <contains [ v]> and <is [ v]> and similar blocks are already in operators section.

However, I would agree adding them as an extension if there would be added along with other new blocks in extension like "string tools".

> You can find out more about extensions in the VM, but that feature is still getting worked on

When do you expect support for extensions to be finished?

**Comment by cwillisf** [comment]
This is a feature request for the Scratch language or runtime rather than for the scratch-blocks editor. scratch-blocks handles the visual block editor; changes to the block library, language, or runtime would be tracked elsewhere. Closing as part of an issue cleanup. Thank you for the suggestion!

If you believe this issue is still relevant to scratch-blocks v2, please let us know and we can reopen it. It may also be appropriate to open a new issue in the [scratch-editor](https://github.com/scratchfoundation/scratch-editor) repository instead.
